pkg/reconciler/pipeline/controller_test.go

```diff
@@ -144,6 +146,60 @@ func Test_reconcile(t *testing.T) {
 	})
 }
 
+func Test_reconcileEvents(t *testing.T) {
+	t.Run("test reconcile - invalid name", func(t *testing.T) {
+		cl := fake.NewClientBuilder().Build()
+		ctx := context.TODO()
+		testIsbSvc := testNativeRedisIsbSvc.DeepCopy()
+		testIsbSvc.Status.MarkConfigured()
+		testIsbSvc.Status.MarkDeployed()
+		err := cl.Create(ctx, testIsbSvc)
+		assert.Nil(t, err)
+		r := &pipelineReconciler{
+			client:   cl,
+			scheme:   scheme.Scheme,
+			config:   fakeConfig,
+			image:    testFlowImage,
+			logger:   zaptest.NewLogger(t).Sugar(),
+			recorder: record.NewFakeRecorder(64),
+		}
+		testObj := testPipeline.DeepCopy()
+		_, err = r.reconcile(ctx, testObj)
+		assert.NoError(t, err)
+		testObj.Name = "very-very-very-loooooooooooooooooooooooooooooooooooong"
+		_, err = r.reconcile(ctx, testObj)
+		assert.Error(t, err)
+		events := getEvents(r, 1)
+		assert.Equal(t, "Warning ReconcilePipelineFailed the length of the pipeline name plus the vertex name is over the max limit. (very-very-very-loooooooooooooooooooooooooooooooooooong-input), [must be no more than 63 characters]", events[0])
+	})
+
+	t.Run("test reconcile - duplicate vertex", func(t *testing.T) {
+		cl := fake.NewClientBuilder().Build()
+		ctx := context.TODO()
+		testIsbSvc := testNativeRedisIsbSvc.DeepCopy()
+		testIsbSvc.Status.MarkConfigured()
+		testIsbSvc.Status.MarkDeployed()
+		err := cl.Create(ctx, testIsbSvc)
+		assert.Nil(t, err)
+		r := &pipelineReconciler{
+			client:   cl,
+			scheme:   scheme.Scheme,
+			config:   fakeConfig,
+			image:    testFlowImage,
+			logger:   zaptest.NewLogger(t).Sugar(),
+			recorder: record.NewFakeRecorder(64),
+		}
+		testObj := testPipeline.DeepCopy()
+		_, err = r.reconcile(ctx, testObj)
+		assert.NoError(t, err)
+		testObj.Spec.Vertices = append(testObj.Spec.Vertices, dfv1.AbstractVertex{Name: "input", Source: &dfv1.Source{}})
+		_, err = r.reconcile(ctx, testObj)
+		assert.Error(t, err)
+		events := getEvents(r, 1)
```

Comment:
Can we also add a unit test that emit multiple events within on reconciliation? e.g. a normal event followed by a warning event?

Reply:
Currently, there would need to be multiple reconciliation calls now to get a normal event followed by a warning event as when the NormalEvent for updating a pipeline phase occurs, there is a return statement. Otherwise, we do not hit the check for any nonLifeCycleChanges that would cause a warning.